htlcswitch/link.go

```diff
@@ -1307,35 +1308,49 @@ func (l *channelLink) handleDownStreamPkt(pkt *htlcPacket, isReProcess bool) {
 					},
 				)
 
-				// Encrypt the error back to the source unless
-				// the payment was generated locally.
+				// If the payment was locally initiated (which
+				// is indicated by a nil obfuscator), we do
+				// not need to encrypt it back to the sender.
 				if pkt.obfuscator == nil {
 					var b bytes.Buffer
 					err := lnwire.EncodeFailure(&b, failure, 0)
 					if err != nil {
-						l.log.Errorf("unable to encode failure: %v", err)
+						l.log.Errorf("unable to "+
+							"encode failure: %v", err)
 						l.mailBox.AckPacket(pkt.inKey())
 						return
 					}
 					reason = lnwire.OpaqueReason(b.Bytes())
-					localFailure = true
 				} else {
+					// If the packet is part of a forward,
+					// (identified by a non-nil obfuscator)
+					// we need to encrypt the error back to
+					// the source.
 					var err error
 					reason, err = pkt.obfuscator.EncryptFirstHop(failure)
 					if err != nil {
-						l.log.Errorf("unable to obfuscate error: %v", err)
+						l.log.Errorf("unable to "+
+							"obfuscate error: %v", err)
 						l.mailBox.AckPacket(pkt.inKey())
 						return
 					}
 				}
 
+				// create a link error containing the temporary
+				// channel failure and a detail which indicates
+				// the we failed to add the htlc.
+				linkError := NewDetailedLinkError(
+					failure,
+					FailureDetailDownstreamHtlcAdd,
+				)
+
 				failPkt := &htlcPacket{
 					incomingChanID: pkt.incomingChanID,
 					incomingHTLCID: pkt.incomingHTLCID,
 					circuit:        pkt.circuit,
 					sourceRef:      pkt.sourceRef,
 					hasSource:      true,
-					localFailure:   localFailure,
+					linkFailure:    linkError,
```

Comment:
I think this is a more natural way to think about this information:
One of our local payments failed at our node-> when handling local errors, don't encrypt
vs
This htlc failed at our node -> if it is local, do not encrypt the error

We are relying on local payments being handled by the special case `handleLocalResponse`, which is not a change to our current set of assumptions? I do think that this change clarifies the information flow, but I do see the concern of adding more implicit knowledge to understanding the switch, so I will think about this one more.

Reply:
No, the special case `handleLocalResponse` is nothing new. 

I am thinking about the assumption that if `linkFailure` is set, the reason is not encrypted. This is not always true, but not a problem because of how we use the result. And then there is also that `convertedError` boolean to signal that the reason is not encrypted, even though there is no `linkFailure` set. See https://github.com/lightningnetwork/lnd/pull/3027#discussion_r280013085 for discussion about this short cut.

What may also contribute to the confusion, is that `CheckHtlcForward` returns a failure for which it lets the switch encrypt the reason and create the fail packet. It isn't clear which unit is responsible for encrypting failures. Maybe eventually the switch needs to always do the encryption, to clearly separate the responsibilities. The link can then always return unencrypted failures and we'd also not need this implicit and explicit signaling of the state of the failure reason.

Reply:
fwiw I think this change makes it clearer what is going on than the boolean. If we just document that the reason can be encrypted even though `linkFailure` is set, then there is no confusion?

I agree that the responsibilities aren't that nicely separated here, like relying on the obfuscator being nil to determine whether this is a local payment. This could possibly be improved by having a no-op obfuscator doing the encrypting for local payment, or doing the separating as @joostjager is suggesting, but both seem like much bigger changes than we would want here.

Reply:
> I am thinking about the assumption that if linkFailure is set, the reason is not encrypted.

I don't think that this change is putting that assumption in place? If `linkFailure` is set, the assumption is that the htlc failed at our node. Using this field in isolation to determine whether something is encrypted is definitely wrong. The sticking point for me here is that we are not basing this logic only on the value of `linkFailure`, we are pairing this information with the fact that we have a local payment: 
Incorrect assumption: `linkFailure!=nil -> unencrypted`
Acceptable assumption: `if local && linkFailure!=nil -> unencrypted`   

>  And then there is also that convertedError boolean to signal that the reason is not encrypted, even though there is no linkFailure set. 

From my understanding of convertedError, it covers the case where we receive `UpdateFailMalformedHTLC`? A nil link error makes sense to me here because we didn't fail the htlc on one of our links. It failed further down the route, this failure just happens to be a special one where the peer can't parse the htlc so we need to encrypt an error for them. 

If we are more comfortable to touch as little of the existing logic possible, I can put the bool back and open an issue with some of the discussion from this PR/the one you linked @joostjager? Because doing this properly is definitely a bigger refactor than the scope of this PR.

Reply:
I don't mean to say that the code isn't correct, but just that the packet isn't self-describing whether the reason is encrypted or not. Additional context is needed to interpret the field. I think that that isn't ideal, especially because there is already so much going on. 

I'd prefer to have an explicit bool saying what state `reason` is in. Maybe that boolean can be `convertedError` renamed to `isEncrypted` even. The `linkFailure` can be in the packet too for the failed adds.

Fixing the shortcut itself mentioned above (checking the length) is definitely out of scope for this PR. My concern is just about communicating the encrypted status back to the switch.

Reply:
> Additional context is needed to interpret the field. I think that that isn't ideal, especially because there is already so much going on.

Ok got it, moving away from requiring additional context rather than adding is going to be better long term. I thought it was fine because we already do it, but I can see that it'll just add extra cleanup for the future. 

> I'd prefer to have an explicit bool saying what state reason is in. Maybe that boolean can be convertedError renamed to isEncrypted even. 

This makes sense, will update to have one more explanatory bool. `localPaymentErrorEncryted` too long?

Reply:
Hm actually, `convertedError` needs to be distinguished from `localFailure` [here](https://github.com/lightningnetwork/lnd/blob/master/htlcswitch/switch.go#L1116) otherwise a unified bool is just getting us back into the same contextual field issue. I think I'm just going to not touch this for now, and rather come back to it with a full focus on refactoring when needed rather than make small semi-refactors.